lang/src/org/partiql/lang/eval/builtins/DynamicLookupExprFunction.kt

```diff
@@ -0,0 +1,101 @@
+package org.partiql.lang.eval.builtins
+
+import org.partiql.lang.errors.ErrorCode
+import org.partiql.lang.eval.BindingCase
+import org.partiql.lang.eval.BindingName
+import org.partiql.lang.eval.EvaluationException
+import org.partiql.lang.eval.EvaluationSession
+import org.partiql.lang.eval.ExprFunction
+import org.partiql.lang.eval.ExprValue
+import org.partiql.lang.eval.ExprValueType
+import org.partiql.lang.eval.physical.throwUndefinedVariableException
+import org.partiql.lang.eval.stringValue
+import org.partiql.lang.types.FunctionSignature
+import org.partiql.lang.types.StaticType
+import org.partiql.lang.types.VarargFormalParameter
+
+/**
+ * Performs dynamic variable resolution.  Query authors should never call this function directly (and indeed it is
+ * named to avoid collision with the names of custom functions)--instead, the query planner injects call sites
+ * to this function to perform dynamic variable resolution of undefined variables.  This provides a migration path
+ * for legacy customers that depend on this behavior.
+ *
+ * Arguments:
+ *
+ * - variable name
+ * - case sensitivity
+ * - lookup strategy (globals then locals or locals then globals)
+ * - A variadic list of locations to be searched.
+ *
+ * The variadic arguments must be of type `any` because the planner doesn't yet have knowledge of static types
+ * and therefore cannot filter out local variables types that are not structs.
+ */
+class DynamicLookupExprFunction : ExprFunction {
```

Comment:
Oops, this should have been introduced in a later PR.  I'll look at moving this to one of the later PRs.